src/lang/std/engineConnection.ts

```diff
@@ -48,17 +48,94 @@ type Timeout = ReturnType<typeof setTimeout>
 
 type ClientMetrics = Models['ClientMetrics_type']
 
+type Value<T, U> = U extends undefined
+  ? { type: T; value: U }
+  : U extends void
+  ? { type: T }
+  : { type: T; value: U }
+
+enum EngineConnectionStateType {
+  Fresh = 'fresh', // Don't want to use New since it's usually a keyword
+  Connecting = 'connecting',
+  ConnectionEstablished = 'connection-established',
+  Disconnected = 'disconnected',
+}
+
+enum DisconnectedType {
+  Error = 'error',
+  Timeout = 'timeout',
+  Quit = 'quit',
+}
+
+type DisconnectedValue =
+  | Value<DisconnectedType.Error, Error | undefined>
+  | Value<DisconnectedType.Timeout, void>
+  | Value<DisconnectedType.Quit, void>
+
+enum ConnectingType {
+  WebSocketConnecting = 'websocket-connecting',
+  WebSocketEstablished = 'websocket-established',
+  PeerConnectionCreated = 'peer-connection-created',
+  ICEServersSet = 'ice-servers-set',
+  SetLocalDescription = 'set-local-description',
+  OfferedSdp = 'offered-sdp',
+  SetRemoteDescription = 'set-remote-description',
+  WebRTCConnecting = 'webrtc-connecting',
+  TrackReceived = 'track-received',
+  ICECandidateReceived = 'ice-candidate-received',
+  DataChannelConnecting = 'data-channel-connecting',
+  DataChannelEstablished = 'data-channel-established',
+}
+
+type ConnectingValue =
+  | Value<ConnectingType.WebSocketConnecting, void>
+  | Value<ConnectingType.WebSocketEstablished, void>
+  | Value<ConnectingType.PeerConnectionCreated, void>
+  | Value<ConnectingType.ICEServersSet, void>
+  | Value<ConnectingType.SetLocalDescription, void>
+  | Value<ConnectingType.OfferedSdp, void>
+  | Value<ConnectingType.SetRemoteDescription, void>
+  | Value<ConnectingType.WebRTCConnecting, void>
+  | Value<ConnectingType.TrackReceived, void>
+  | Value<ConnectingType.ICECandidateReceived, void>
+  | Value<ConnectingType.DataChannelConnecting, string>
+  | Value<ConnectingType.DataChannelEstablished, void>
+
+type EngineConnectionState =
+  | Value<EngineConnectionStateType.Fresh, void>
```

Comment:
Great idea - I can't think of a pragmatic example where we need to differentiate between `Fresh` and `Offline` so it makes sense one can be the other, but the latter is just a better name :sunglasses:

Reply:
Nope, I just encountered why this is an issue.

Offline doesn't tell us if we've already connected once or not at all. Fresh, and Disconnected, tell us if we're at the beginning or end of a connection, and both together imply "Offline".

Reply:
Gotcha. It would be nice to have an additional "Offline" state perhaps, where if the browser is actually offline we transition from `Fresh` -> `Offline` so that the network indicator can pick that up -- it'd let us remove quite a bit of code, and I figure it can't hurt since we have to be network aware here too. Any objections?

Reply:
My only objection is Offline and Disconnect are really similar on the surface :s

Reply:
Yeah I'm with you, I was trying to smooshify one of the states with offline in my head and couldn't figure out which one made the most sense. Offline is also one we can transition from if `window.addEventListener('online', ...)` fires, and we may transition abruptly to Offline if `window.addEventListener('offline', ..)` fires.

My only real motivation with this line of thinking is that I'd really /really/ love to remove the code from the network indicator react component thinger if it's possible, even at the expense of a bit of conceptual clarity we have to comment our way through -- doubly so since I want to have the network indicator show more than just "online/offline", likely by reading the `state` attribute off the connection here, so it's a bit easier to see why the connection is offline -- I don't have any strong opinions or motivations aside from that integration

Reply:
If it helps, Offline would just be defined as `state == (Fresh or Disconnected)` :p

Reply:
We're going to defer thinking about this to a new PR. I think I know what to do, but its going to involve some more work, and I don't want to continue growing the scope of the PR even if directly related